topiary/src/tree_sitter.rs

```diff
@@ -225,34 +224,82 @@ fn collect_leaf_ids(matches: &Vec<LocalQueryMatch>, capture_names: &[String]) ->
     ids
 }
 
-// TODO: Deduplicate these.
-
-fn handle_delimiter_predicate(predicate: &QueryPredicate) -> FormatterResult<Option<String>> {
+fn handle_predicate(
+    predicate: &QueryPredicate,
+    predicates: &QueryPredicates,
+) -> FormatterResult<QueryPredicates> {
     let operator = &*predicate.operator();
-
     if let "delimiter!" = operator {
         let arg =
             predicate.args().into_iter().next().ok_or_else(|| {
                 FormatterError::Query(format!("{operator} needs an argument"), None)
             })?;
-
-        Ok(Some(arg))
-    } else {
-        Ok(None)
-    }
-}
-
-fn handle_scope_id_predicate(predicate: &QueryPredicate) -> FormatterResult<Option<String>> {
-    let operator = &*predicate.operator();
-
-    if let "scope_id!" = operator {
+        Ok(QueryPredicates {
+            delimiter: Some(arg),
+            ..predicates.clone()
+        })
+    } else if let "scope_id!" = operator {
         let arg =
             predicate.args().into_iter().next().ok_or_else(|| {
                 FormatterError::Query(format!("{operator} needs an argument"), None)
             })?;
+        Ok(QueryPredicates {
+            scope_id: Some(arg),
+            ..predicates.clone()
```

Comment:
All the clones here look suspect, but I've not studied the code enough to see if you really need them. Have you considered if there is a better way?